### ISSUE TITLE
Bug: (regression) cooperative grid sync w/ cache=True - working in 0.58.1, not working in 0.59.0

### ISSUE THREAD

**Issue by shaunc** [numba/numba#9432]
## Reporting a bug

Ability to cache kernels using cooperative grid sync was added in https://github.com/numba/numba/pull/8895.

Unfortunately this seems to be broken again in 0.59.0, though is working in 0.58.1.

Test:
```py
from numba import cuda  # type: ignore


@cuda.jit(cache=True)  # type: ignore
def assign_edges_kernel() -> None:
    """
    Kernel which tests the assign_edges function.
    """
    gg = cuda.cg.this_grid()  # type: ignore
    gg.sync()


if __name__ == "__main__":
    assign_edges_kernel[1, 1]()  # type: ignore
```

Result in 0.58.1:
```
$ poetry add numba==0.58.1

Updating dependencies
Resolving dependencies... (0.4s)

Package operations: 0 installs, 2 updates, 0 removals

  • Downgrading llvmlite (0.42.0 -> 0.41.1)
  • Downgrading numba (0.59.0 -> 0.58.1)

Writing lock file
(python3.11) shauncutts@silk:~/.../mlops/ingolstadt$ poetry run python3.11 test.py
/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/dispatcher.py:536: NumbaPerformanceWarning: Grid size 1 will likely result in GPU under-utilization due to low occupancy.
  warn(NumbaPerformanceWarning(msg))
  ```
  
  Result in 0.59.0:
  ```
  $ poetry add numba==0.59.0

Updating dependencies
Resolving dependencies... (0.4s)

Package operations: 0 installs, 2 updates, 0 removals

  • Updating llvmlite (0.41.1 -> 0.42.0)
  • Updating numba (0.58.1 -> 0.59.0)

Writing lock file
(python3.11) shauncutts@silk:~/.../mlops/ingolstadt$ poetry run python3.11 test.py
/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/dispatcher.py:536: NumbaPerformanceWarning: Grid size 1 will likely result in GPU under-utilization due to low occupancy.
  warn(NumbaPerformanceWarning(msg))
Traceback (most recent call last):
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/errors.py", line 860, in new_error_context
    yield
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 270, in lower_block
    self.lower_inst(inst)
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 448, in lower_inst
    val = self.lower_assign(ty, inst)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 660, in lower_assign
    return self.lower_expr(ty, value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 1196, in lower_expr
    res = self.lower_call(resty, expr)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 925, in lower_call
    res = self._lower_call_normal(fnty, expr, signature)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 1159, in _lower_call_normal
    impl = self.context.get_function(fnty, signature)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/base.py", line 556, in get_function
    return self.get_function(fn, sig, _firstcall=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/base.py", line 558, in get_function
    raise NotImplementedError("No definition for lowering %s%s" % (key, sig))
NotImplementedError: No definition for lowering (<class 'numba.cuda.types.GridGroup'>, 'sync')(GridGroup,) -> int32

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/shauncutts/src/factfiber.ai/mlops/ingolstadt/test.py", line 14, in <module>
    assign_edges_kernel[1, 1]()  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/dispatcher.py", line 539, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/dispatcher.py", line 673, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/dispatcher.py", line 681, in _compile_for_args
    return self.compile(tuple(argtypes))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/dispatcher.py", line 924, in compile
    kernel = _Kernel(self.py_func, argtypes, **self.targetoptions)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/dispatcher.py", line 83, in __init__
    cres = compile_cuda(self.py_func, types.void, self.argtypes,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/cuda/compiler.py", line 196, in compile_cuda
    cres = compiler.compile_extra(typingctx=typingctx,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler.py", line 751, in compile_extra
    return pipeline.compile_extra(func)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler.py", line 445, in compile_extra
    return self._compile_bytecode()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler.py", line 513, in _compile_bytecode
    return self._compile_core()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler.py", line 492, in _compile_core
    raise e
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler.py", line 479, in _compile_core
    pm.run(self.state)
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 368, in run
    raise patched_exception
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 356, in run
    self._runPass(idx, pass_inst, state)
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 311, in _runPass
    mutated |= check(pss.run_pass, internal_state)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 273, in check
    mangled = func(compiler_state)
              ^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/typed_passes.py", line 468, in run_pass
    lower.lower()
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 187, in lower
    self.lower_normal_function(self.fndesc)
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 226, in lower_normal_function
    entry_block_tail = self.lower_function_body()
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 256, in lower_function_body
    self.lower_block(block)
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/lowering.py", line 268, in lower_block
    with new_error_context('lowering "{inst}" at {loc}', inst=inst,
  File "/usr/local/share/mambaforge/envs/python3.11/lib/python3.11/contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/shauncutts/.cache/pypoetry/virtualenvs/ingolstadt-EH7L9v-Y-py3.11/lib/python3.11/site-packages/numba/core/errors.py", line 874, in new_error_context
    raise newerr.with_traceback(tb)
numba.core.errors.LoweringError: Failed in cuda mode pipeline (step: native lowering)
No definition for lowering (<class 'numba.cuda.types.GridGroup'>, 'sync')(GridGroup,) -> int32

File "test.py", line 10:
def assign_edges_kernel() -> None:
    <source elided>
    gg = cuda.cg.this_grid()  # type: ignore
    gg.sync()
    ^

During: lowering "$92call.8 = call $66load_method.7(func=$66load_method.7, args=[], kws=(), vararg=None, varkwarg=None, target=None)" at /home/shauncutts/src/factfiber.ai/mlops/ingolstadt/test.py (10)
```

**Comment by gmarkall** [comment]
Thanks for the report - we actually have a unit test for exactly this case, but it turns out to have been spuriously passing. If one runs:

```
python runtests.py numba.cuda.tests.cudapy.test_caching -v -f
```

then we get a failure as expected:

```
test_cache (numba.cuda.tests.cudapy.test_caching.TestMultiCCCaching.test_cache) ... skipped 'Need two different CCs for multi-CC cache test'
test_cache_cg (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_cache_cg) ... ERROR

======================================================================
ERROR: test_cache_cg (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_cache_cg)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_caching.py", line 167, in test_cache_cg
    mod.cg_usecase(0)
  File "/tmp/numba-tests.1000/test_cache-er4vyx0f/cuda_caching_test_fodder.py", line 31, in __call__
    self._call(array_return, *array_args)
  File "/tmp/numba-tests.1000/test_cache-er4vyx0f/cuda_caching_test_fodder.py", line 41, in _call
    self._func[1, 1](ret, *args)
  File "/home/gmarkall/numbadev/numba/numba/cuda/dispatcher.py", line 539, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/cuda/dispatcher.py", line 673, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/cuda/dispatcher.py", line 681, in _compile_for_args
    return self.compile(tuple(argtypes))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/cuda/dispatcher.py", line 924, in compile
    kernel = _Kernel(self.py_func, argtypes, **self.targetoptions)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/cuda/dispatcher.py", line 83, in __init__
    cres = compile_cuda(self.py_func, types.void, self.argtypes,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 196, in compile_cuda
    cres = compiler.compile_extra(typingctx=typingctx,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 751, in compile_extra
    return pipeline.compile_extra(func)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 445, in compile_extra
    return self._compile_bytecode()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 513, in _compile_bytecode
    return self._compile_core()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 488, in _compile_core
    raise e
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 479, in _compile_core
    pm.run(self.state)
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 364, in run
    raise e
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 356, in run
    self._runPass(idx, pass_inst, state)
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 311, in _runPass
    mutated |= check(pss.run_pass, internal_state)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 273, in check
    mangled = func(compiler_state)
              ^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/typed_passes.py", line 468, in run_pass
    lower.lower()
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 187, in lower
    self.lower_normal_function(self.fndesc)
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 226, in lower_normal_function
    entry_block_tail = self.lower_function_body()
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 256, in lower_function_body
    self.lower_block(block)
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 270, in lower_block
    self.lower_inst(inst)
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 448, in lower_inst
    val = self.lower_assign(ty, inst)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 660, in lower_assign
    return self.lower_expr(ty, value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 1196, in lower_expr
    res = self.lower_call(resty, expr)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 925, in lower_call
    res = self._lower_call_normal(fnty, expr, signature)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 1159, in _lower_call_normal
    impl = self.context.get_function(fnty, signature)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/base.py", line 556, in get_function
    return self.get_function(fn, sig, _firstcall=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/numbadev/numba/numba/core/base.py", line 558, in get_function
    raise NotImplementedError("No definition for lowering %s%s" % (key, sig))
NotImplementedError: No definition for lowering (<class 'numba.cuda.types.GridGroup'>, 'sync')(GridGroup,) -> int32

----------------------------------------------------------------------
Ran 2 tests in 0.147s

FAILED (errors=1, skipped=1)
```

However if we also run `test_compiler`, then we get a spurious pass:

```
python runtests.py numba.cuda.tests.cudapy.test_compiler numba.cuda.tests.cudapy.test_caching -v -f
```

gives 

```
test_c_abi_device_function (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_c_abi_device_function) ... ok
test_c_abi_device_function_module_scope (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_c_abi_device_function_module_scope) ... ok
test_c_abi_disallowed_for_kernel (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_c_abi_disallowed_for_kernel) ... ok
test_c_abi_with_abi_name (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_c_abi_with_abi_name) ... ok
test_cache (numba.cuda.tests.cudapy.test_caching.TestMultiCCCaching.test_cache) ... skipped 'Need two different CCs for multi-CC cache test'
test_cache_cg (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_cache_cg) ... ok
test_cache_invalidate (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_cache_invalidate) ... ok
test_cache_reuse (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_cache_reuse) ... ok
test_caching (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_caching) ... ok
test_cannot_cache_linking_libraries (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_cannot_cache_linking_libraries) ... ok
test_closure (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_closure) ... ok
test_cpu_and_cuda_reuse (numba.cuda.tests.cudapy.test_caching.CUDAAndCPUCachingTest.test_cpu_and_cuda_reuse) ... ok
test_cpu_and_cuda_targets (numba.cuda.tests.cudapy.test_caching.CUDAAndCPUCachingTest.test_cpu_and_cuda_targets) ... ok
test_device_function (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_device_function) ... ok
test_device_function_with_debug (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_device_function_with_debug) ... /home/gmarkall/numbadev/numba/numba/cuda/compiler.py:303: NumbaInvalidConfigWarning: debug=True with opt=True (the default) is not supported by CUDA. This may result in a crash - set debug=False or opt=False.
  warn(NumbaInvalidConfigWarning(msg))
ok
test_device_function_with_line_info (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_device_function_with_line_info) ... ok
test_fastmath (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_fastmath) ... ok
test_global_kernel (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_global_kernel) ... ok
test_kernel_with_debug (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_kernel_with_debug) ... /home/gmarkall/numbadev/numba/numba/cuda/compiler.py:303: NumbaInvalidConfigWarning: debug=True with opt=True (the default) is not supported by CUDA. This may result in a crash - set debug=False or opt=False.
  warn(NumbaInvalidConfigWarning(msg))
ok
test_kernel_with_line_info (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_kernel_with_line_info) ... ok
test_many_locals (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_many_locals) ... ok
test_multiprocessing (numba.cuda.tests.cudapy.test_caching.TestMultiprocessCache.test_multiprocessing) ... ok
test_nanosleep (numba.cuda.tests.cudapy.test_compiler.TestCompileOnlyTests.test_nanosleep) ... ok
test_no_caching (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_no_caching) ... ok
test_non_creatable_pycache (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_non_creatable_pycache) ... ok
test_non_void_return_type (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_non_void_return_type) ... ok
test_non_writable_pycache (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_non_writable_pycache) ... ok
test_recompile (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_recompile) ... ok
test_same_names (numba.cuda.tests.cudapy.test_caching.CUDACachingTest.test_same_names) ... ok
test_unsupported_abi (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTX.test_unsupported_abi) ... ok
test_cannot_serialize_unfinalized (numba.cuda.tests.cudapy.test_caching.TestCUDACodeLibrary.test_cannot_serialize_unfinalized) ... ok
test_compile_ptx_for_current_device (numba.cuda.tests.cudapy.test_compiler.TestCompileToPTXForCurrentDevice.test_compile_ptx_for_current_device) ... ok

----------------------------------------------------------------------
Ran 32 tests in 5.331s

OK (skipped=1)
```

which is what we've been seeing in CI all along (the test always passing, even though the functionality doesn't generally work).

**Comment by gmarkall** [comment]
#9447 contains a fix for this and we'll aim to get it in 0.59.1.

@shaunc in the meantime as a workaround, I think you can just compile an empty uncached kernel first, and then the targets should get initialized correcltly, e.g.:

```python
from numba import cuda  # type: ignore

# Force initialization of CUDA target prior to cache check with a compilation
cuda.jit('void()')(lambda: None)


@cuda.jit(cache=True)  # type: ignore
def assign_edges_kernel() -> None:
    """
    Kernel which tests the assign_edges function.
    """
    gg = cuda.cg.this_grid()  # type: ignore
    gg.sync()


if __name__ == "__main__":
    assign_edges_kernel[1, 1]()  # type: ignore
```

which executes without error for me.

**Comment by shaunc** [comment]
Thank you!
